### ISSUE TITLE
Binding Autocomplete doesn't work

### ISSUE THREAD

**Issue by arnederuwe** [aurelia/vscode-extension#116]
**I'm submitting a bug report**

* **Extension Version:**
1.0.5

**Please tell us about your environment:**
* **Operating System:**
Windows 10

* **Visual studio code version:**
1.41.1

**Current behavior:**
![image](https://user-images.githubusercontent.com/3369747/72061689-af14fd80-32d6-11ea-85f7-7c856687727b.png)
I don't get the properties from my backing viewmodel in my view intellisense. My Aurelia plugin settings are defined like this:
```
	"aurelia.featureToggles": {
		"smartAutocomplete": true
	},
```

**Expected/desired behavior:**

* **What is the expected behavior?**
I would expect it to behave in the way described [here](https://aurelia.io/blog/2020/01/01/aurelia-release-notes-january-2020) 

I suspect my plugin isn't properly configured, as almost all other functionalities in that blogpost aren't working as well. The only thing that does work is the open related files functionality.

**Comment by hiaux0** [comment]
Could you try and see, if functionality becomes available, after you launched this command?:

![image](https://user-images.githubusercontent.com/30693990/72062787-c1903680-32d8-11ea-825b-96438b879136.png)

`Show Aurelia view data to side`

At least, it should open a web view window in vscode, hopefully.

**Comment by hiaux0** [comment]
> I don't get the properties from my backing viewmodel in my view intellisense. My Aurelia plugin settings are defined like this:

Could you try to delete the quotes and retype them?

![gifgif](https://user-images.githubusercontent.com/30693990/72063265-c0abd480-32d9-11ea-921c-c1a06da197f3.gif)

**Comment by arnederuwe** [comment]
Hi, thanks for the quick reply!

Deleting the quotes and retyping them doesn't fix the issue unfortunately.

Typing your command opens a tab with the following info:
Component: 'app'
Files
c:/Projects/Playground/plugin-broken-test/src/app.html
c:/Projects/Playground/plugin-broken-test/src/app.ts
ViewModel
type: typescript

Properties
message (string)
Methods
test (undefined) => (params: )
View
exports to this view
App

**Comment by hiaux0** [comment]
Will try to setup the extension with a fresh vscode installation on another machine.

**Comment by arnederuwe** [comment]
Fyi, I just tried it again in a [Windows Sandbox](https://www.windowscentral.com/how-use-windows-sandbox-windows-10-may-2019-update) and got the same issue. I installed npm, vscode, the plugin and the aurelia CLI

**Comment by BjoernHund** [comment]
We've also the same problem. AutoComplete doesn't work.

If I try to open Aurelia view data to side, I get the following message:

```No file path found for {Path of the HTML Template)```

The command `Open Related Aurelia File: View` works.

**Comment by arnederuwe** [comment]
Please let me know if there is anything more I can provide in order to get this fixed. It has to be something trivial, as our entire team (6+ people) can't get this to work on all our Aurelia projects. I would be happy to organize a screen share if that would help you.

**Comment by pkkummermo** [comment]
@arnederuwe I think @hiaux0 is working hard on the [PR](https://github.com/aurelia/vscode-extension/pull/117) as we speak :)

**Comment by arnederuwe** [comment]
ahh, I completely missed that.. My apologies :)
@hiaux0, we are using Windows 10 exclusively, if I could lighten your burden and perform some tests on my machines, please don't hesitate to contact me.

**Comment by pkkummermo** [comment]
I have a Win10 setup with WSL Ubuntu18.04 along with VSCode if you need that permutation tested. Just ping me as well :)

**Comment by hiaux0** [comment]
> Please let me know if there is anything more I can provide in order to get this fixed. It has to be something trivial, as our entire team (6+ people) can't get this to work on all our Aurelia projects. I would be happy to organize a screen share if that would help you.

You could try to build #117 (potentially have to checkout the indiviual commits). Then copy paste into extension folder.

Are you on the aurelia [discord](https://discord.gg/HBtxRD)?
We could chat over there.

**Comment by HymanZHAN** [comment]
@hiaux0 Any updates on this? Sorry I am not very familiar with VS Code extension development. May I ask how to build and test locally? I am also on Linux (Fedora 31) if that helps.

**Comment by pkkummermo** [comment]
@hiaux0 just tested the branch on win10 with an existing project (which worked before), and it's now way more consistent and works much better.

**Comment by hiaux0** [comment]
Hey guys, finally some update from my side.
While I still worked on the extension now and then, I could provide you with good steps to test the raw branch out:

##### Here it is!
@HymanZHAN @pkkummermo @arnederuwe 

1. Clone Rep
2. checkout to `fix/autocomplete`
3. Go to the debug section
   1. and run the `client`
   2. --> Should open separate VSCode window, with latest changes included
4. From here, you can open any Aurelia project, and test the latest changes

I also streamed the whole process, and you can follow how I did the steps above:
https://www.twitch.tv/videos/573788060?t=0h29m42s

### Changelog until here:
- Added configuration setting, that hopefully also allows Monorepos (not tested)
```json
	"aurelia.extensionSettings": {
		"pathToAureliaProject": [
			"src", // this is the default
			"some/other/path/one",
			"some/other/path/two",
			"some/other/path/three",
		]
	},
```

The extension, will parse all of these paths

Eg. following folder structure should work

Project
- /src
- /some
---- /other
-------- /path
------------ /one
------------ /two
------------ /three

Note: just specifying "some" would have been enough. (similar to why "src" enough)

**Comment by hiaux0** [comment]
I also added a Wiki

https://github.com/aurelia/vscode-extension/wiki/View-Autocomplete

**Comment by HymanZHAN** [comment]
@hiaux0 Thank you for the update! I manually copied the repo folder with your `fix/autocomplete` branch into `~\.vscode\extensions` (I have been using Win10 recently) and used it and I can confirm that binding works for my first few tries.

However, autocompletion stops working if there are parsing errors when I load the project for the first time.

For example, I am following the todo quick start tutorial so my `app.ts` is:

```typescript
interface Todo {
  description: string;
  done: boolean;
}

export class App {
  heading = "Todos";
  todos: Todo[] = [];
  todoDescription = "";

  addTodo() {
    if (this.todoDescription) {
      this.todos.push({
        description: this.todoDescription,
        done: false
      });
      this.todoDescription = "";
    }
  }

  removeTodo(todo: Todo) {
    let index = this.todos.indexOf(todo);
    if (index !== -1) {
      this.todos.splice(index, 1);
    }
  }
}
```

And my `app.html` is:

```html
<template>
  <h1>${heading}</h1>
  <form submit.trigger="addTodo()">
    <input type="text" value.bind="todoDescription" />
    <button type="submit" disabled.bind="!todoDescription">Add Todo</button>
  </form>

  <ul>
    <li repeat.for="todo of todos">
      <input type="checkbox" checked.bind="todo.done" />
      <span css="text-decoration: ${todo.done ? 'line-through' : 'none'}">
        ${todo.description}
      </span>
      <button click.trigger="">Remove</button>
    </li>
  </ul>
</template>
```

If I load the project with an unfinished template with an unfinished `repeat.for`:

```html
<template>
  <h1>${heading}</h1>
  <form submit.trigger="addTodo()">
    <input type="text" value.bind="todoDescription" />
    <button type="submit" disabled.bind="!todoDescription">Add Todo</button>
  </form>

  <ul>
    <!-- Instead of <li repeat.for="todo of todos"> I have: -->
    <li repeat.for=""> 
      <input type="checkbox" checked.bind="todo.done" />
      <span css="text-decoration: ${todo.done ? 'line-through' : 'none'}">
        ${todo.description}
      </span>
      <button click.trigger="">Remove</button>
    </li>
  </ul>
</template>
```

I will get the following error:

```terminal
Debugger listening on ws://127.0.0.1:6100/1896173f-fbe1-49ae-ab5e-54d1bf6580b5
For help, see: https://nodejs.org/en/docs/inspector
smart auto complete init
>>> 1.1 This is the project's directory:
c:\Users\Xucong\Development\MOOC\learn_aurelia\todo
>>> 1.2. The extension will try to parse Aurelia components inside:
src
>>> 1.3. Eg.
c:\Users\Xucong\Development\MOOC\learn_aurelia\todo/src
failed to parse path c:/Users/Xucong/Development/MOOC/learn_aurelia/todo/src/app.html
"Parser Error: Invalid start of expression at column 0 in expression []"
"Error: Parser Error: Invalid start of expression at column 0 in expression []\n    at ParserImplementation.err (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\node_modules\\aurelia-binding\\dist\\commonjs\\aurelia-binding.js:2895:11)\n    at ParserImplementation.parseBindingBehavior (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\node_modules\\aurelia-binding\\dist\\commonjs\\aurelia-binding.js:2423:12)\n    at Parser.parse (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\node_modules\\aurelia-binding\\dist\\commonjs\\aurelia-binding.js:2387:80)\n    at AureliaHtmlParser.getAttributeCommands (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\dist\\src\\server\\FileParser\\Parsers\\AureliaHtmlParser.js:80:48)\n    at AureliaHtmlParser.<anonymous> (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\dist\\src\\server\\FileParser\\Parsers\\AureliaHtmlParser.js:35:46)\n    at Generator.next (<anonymous>)\n    at fulfilled (c:\\Users\\Xucong\\Development\\Packages\\vscode-extension\\dist\\src\\server\\FileParser\\Parsers\\AureliaHtmlParser.js:4:58)\n    at processTicksAndRejections (internal/process/task_queues.js:85:5)"
>>> 2. The extension found this many components:
3
```

And autocompletion will not work until I have a correct template and then reload the window.

**Comment by hiaux0** [comment]
## Parsing error "... of ..."
Currently, I only know of this "fix"

1. Invoke command: `Open Extensions Folder`
2. Navigate to
`/Users/hdn/.vscode/extensions/aureliaeffect.aurelia-1.0.5/node_modules/aurelia-binding/dist/commonjs/aurelia-binding.js
3. At around line 3026, add the following row
```js
KeywordLookup.in = T$InKeyword;
KeywordLookup.of = T$InKeyword; // <-- this line
KeywordLookup.instanceof = T$InstanceOfKeyword;
```
4. Reload and parsing error should appear

## Parser Error: Invalid start of expression at column 0 in expression []
This appears, when in your view you have 'invalid' aurelia syntax
```html
 <li repeat.for="">  // <- Quotes should not be empty
```

Hope these help!

**Comment by HymanZHAN** [comment]
@hiaux0 Thanks for the reply! Yes, I am aware of the error and why it appears. My main question is that is there a way to automatically trigger re-parsing? As it currently stands, if I have an incorrect template when I first loaded the project, I need to manually correct it and reload the window for the extension to re-parse the project.

I noticed at the end of the Wiki:

> Refreshing your components.
> Currently, if you add for example a new bindable to your component, the extension does not pick that up. You would need to reload VSCode for that. (Backlog)

I believe they are related or of the same problem, but correct me if I am wrong.

Nonetheless, thank you for making the binding autocomplete work!
